delete MiqExpression#merge_where_clauses

kbrock opened the pull request:
Simplify basic report generation to no longer munge sql

Instead of merging sql into a string and passing that merged string as a `:where_clause` to Rbac, the where clause is just put onto the targets.

Rails already does a great job merging multiple where clauses. So we can delete our custom code and use the standard code.

Alt to: https://github.com/ManageIQ/manageiq/pull/18556

kbrock commented:
^ 
fixed cops

jrafanie commented:
@kbrock do we have adequate testing of more complicated where clauses that previously worked with the custom code to ensure the way rails handles it will work for our use cases?  What's your confidence on this?  Do we need this for rails 5.2, jansa/yes?

kbrock commented:
added a spec ^

@jrafanie My confidence is now high.

I was a little concerned that `find_entries` might return something other than a scope. I checked all classes that defined find_entries and ran reports on all these classes. They all define a scope and add the conditions to the report.

I added a test to ensure both conditions are applied

kbrock commented:
^ fixed cops, and added a `id.to_s` (dumb mistake)

NickLaMuro commented on app/models/miq_report/generator.rb:
```diff
@@ -337,7 +337,6 @@ def generate_basic_results(options = {})
       :filter           => conditions,
       :include_for_find => get_include_for_find_rbac,
       :references       => get_include_rbac,
-      :where_clause     => where_clause,
```
Slightly about this change and why this was needed before, and if it isn't going to break something by removing it.

I have my hands in a bit of this code here (the `:extra_cols` bit below), so I am somewhat familiar with it, but unsure how `:where_clause` ties into `Rbac` and how removing this now when it was being merged above previously as well is going to affect things.

kbrock replied:
I was pretty sure that we could apply the where here, or we can pass it in for it to then be applied for us later

kbrock replied:
I guess this way, this condition gets passed into calc_filtered_ids (if apply_rbac_directly / apply_rbac_through_association

the other way, we get those ids and then apply this condition after the fact.

NickLaMuro replied:
Okay, after looking over this again, I think I had it in my head that you were editing a different method here, but that is _**NOT**_ the case.  So I was just confused when asking this question originally...

So now, instead of defining `where_clause` and passing it as an `rbac_option`, we are just defining it on the `targets` scope, and that should have the same result.  I am cool with that change then.  Thanks!

kbrock commented on spec/models/miq_report/generator_spec.rb:
```diff
@@ -67,6 +67,23 @@
+          :where_clause => ["vms.type = ?", "Vm"],
+          :col_order    => %w[id name host.name vendor]
+        )
+        rpt.generate_table(:userid => @user.userid, :where_clause => {:vendor => "redhat"})
```
This does test different formats (string and hash)

I didn't see the need to pass in more varying options since rails already has those options in their test suite

NickLaMuro commented on spec/models/miq_report/generator_spec.rb:
```diff
@@ -67,6 +67,23 @@
                                                                       "max_trend_value" => 700)
       end
+
+      it "handles a basic report with a where_clause and a where clause passed in options" do
```
Maybe this, as a way to clarify this is dealing with "merging WHERE clauses":

```ruby
it "handles merging WHERE clauses from MiqReport#where_clause and options[:where_clause]"`
```

NickLaMuro replied:
Also, I assume this spec passed prior to the change in `app/models/miq_report/generator.rb`, correct?

kbrock replied:
yes, I wrote that spec to ensure I didn't break anything.

kbrock replied:
nice nick. it does not work before. I'll track this one down

NickLaMuro replied:
Looks like Keenan looked into this and fixed it, so resolving.

kbrock commented on spec/models/miq_report/generator_spec.rb:
```diff
@@ -67,6 +67,23 @@
+        rpt.generate_table(:userid => @user.userid, :where_clause => {:vendor => "redhat"})
+
+        expect(rpt.table.size).to eq(1)
+        expect(rpt.table.first.id.to_s).to eq(vm.id.to_s)
```
a little bit laziness not knowing exactly how the data was stored.
But I would assume that a data table would keep the data all as strings for display. I put in the to_s to follow how I'd implement it

kbrock commented:
updated to 2 commits. a before and an after - both are 🍏

jrafanie commented on spec/lib/miq_expression_spec.rb:
```diff
@@ -2441,45 +2441,6 @@
-      expect(MiqExpression.merge_where_clauses({"vms.id" => 5})).to eq("\"vms\".\"id\" = 5")
-    end
-
-    it "concatinates 2 arrays" do
```
lol, at least we spelled it consistently wrong

jrafanie commented:
please don't forget jansa labels for 5.2 stuff 😉

simaishi commented:
Jansa backport details:

```
$ git log -1
commit d441ba154bc8ec0abe1e536bca113e084b9b23d6
Author: Joe Rafaniello <jrafanie@users.noreply.github.com>
Date:   Tue Apr 14 18:24:44 2020 -0400

    Merge pull request #20052 from kbrock/miq_expression_merge

    delete MiqExpression#merge_where_clauses

    (cherry picked from commit 0b903c2eb44c34b13c9a46c9f5541a7189b4ad28)
```
